tools/get-keys-from-ts.js

```diff
@@ -0,0 +1,546 @@
+                        break;
+                    }
+
+                    return;
+                } default:
+                    return;
+            }
+
+            node[property] = null;
+        });
+
+        traverseExtends(tsDeclarationNode, ({ iName, excluded }) => {
+            const innerTsDeclarationNode = findTsInterfaceDeclaration(iName);
+
+            if (!innerTsDeclarationNode) {
+                unrecognizedTSTypeReferences.add(iName);
+                return;
+            }
+
+            addPropertyToNodeForDeclaration(innerTsDeclarationNode, node, excluded);
+        });
+    }
+
+    for (const tsDeclarationNode of exportedTsInterfaceDeclarations) {
+        const bodyType = tsDeclarationNode.body.body.find(
+            prop => prop.key.name === "type"
+        );
+
+        const typeName = bodyType && bodyType.typeAnnotation &&
+            bodyType.typeAnnotation.typeAnnotation &&
+            bodyType.typeAnnotation.typeAnnotation.literal &&
+            bodyType.typeAnnotation.typeAnnotation.literal.value;
+
+        if (!typeName) {
+            continue;
+        }
+
+        const node = {};
+
+        addPropertyToNodeForDeclaration(tsDeclarationNode, node);
+
+        initialNodes[typeName] = [...new Set(getKeys(node), ...(initialNodes[typeName] || []))];
+    }
+
+    const nodes = alphabetizeKeyInterfaces(initialNodes);
+
+    if (unrecognizedTSTypes.size) {
+        throw new Error(
+            "Unhandled TypeScript type; please update the code to " +
+            "handle the type or if not relevant, add it to " +
+            "`unrecognizedTSTypes`; see\n\n  " +
+            `${[...unrecognizedTSTypes].join(", ")}\n`
+        );
+    }
+    if (unrecognizedTSTypeReferences.size) {
+        throw new Error(
+            "Unhandled TypeScript type reference; please update the code to " +
+            "handle the type reference or if not relevant, add it to " +
+            "`unrecognizedTSTypeReferences`; see\n\n  " +
+            `${[...unrecognizedTSTypeReferences].join(", ")}\n`
+        );
+    }
+
+    return {
+        keys: nodes,
+        tsInterfaceDeclarations: {
+            allTsInterfaceDeclarations,
+            exportedTsInterfaceDeclarations,
+            tsTypeDeclarations
+        }
+    };
+}
+
+/**
+ * @typedef {{tsInterfaceDeclarations: {
+ *   allTsInterfaceDeclarations: {
+ *     Node[],
+ *     keys: KeysStrict
+ *   },
+ *   exportedTsInterfaceDeclarations:
+ *     Node[],
+ *     keys: KeysStrict
+ *   }
+ * }}} VisitorKeysExport
+ */
+
+/**
+ * Builds visitor keys based on TypeScript declaration.
+ * @param {string} file TypeScript declaration file to parse.
+ * @param {{supplementaryDeclarations: Object<string, Node[]>}} options The options
+ * @returns {Promise<VisitorKeysExport>} The built visitor keys
+ */
+async function getKeysFromTsFile(file, options) {
+    const code = await readFile(file);
+
+    return await getKeysFromTs(code, options);
```

Comment:
It technically is, and it comes at the cost of an extra microtask, but as per the [no-return-await](https://eslint.org/docs/rules/no-return-await) rule:

> You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.

Reply:
Yes, but `getKeysFromTs` is no longer `async` so `getKeysFromTs(code, options)` is not a Promise?

Reply:
🤦  Yeah :-)